File: Project/feather1/Assets/Plugins/iOS/XML_StageSaxParser.cpp
#include "XML_StageSaxParser.h"
#include "XML_ProjectLoader.h"

#include "TR_NodeTree.h"
#include "TR_NodeTreeBuilder.h"

#include "RD_ClipDataCore.h"

#include "STD_Types.h"

#include <string.h>

//  Common tags
#define kNameTag               "name"
#define kFilenameTag           "filename"
#define kRepeatTag             "repeat"

//  Stage group main tags
#define kStageColTag           "stages"
#define kStageTag              "stage"
#define kPropColTag            "props"
#define kPlayTag               "play"
#define kPropTag               "prop"
#define kSoundTag              "sound"

//  Stage group variable tags
#define kSkeletonTag           "skeleton"
#define kAnimationTag          "animation"
#define kDrawingAnimationTag   "drawingAnimation"

#define kTimeTag               "time"
#define kParentSkeletonTag     "parent"
#define kParentBoneTag         "bone"

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
#if 0
#pragma mark - XML_StageSaxParser
#endif

XML_StageSaxParser::XML_StageSaxParser( const STD_String &projectFolder, const STD_String &clipName, RD_ClipDataCore *clip ) :
  _projectFolder(projectFolder),
  _clipName(clipName),
  _state(eRoot),
  _clip(clip)
{
  _clip->beginClipData();
}

XML_StageSaxParser::~XML_StageSaxParser()
{
  _clip->endClipData();
}

XML_SaxParserComponentPtr_t XML_StageSaxParser::startComponent(const char *nodeName, const XML_AttributeCol_t &attributes)
{
  if ( _state == eRoot )
  {
    if ( strcmp( nodeName, kStageColTag ) == 0 )
    {
      _state = eStages;
    }
  }
  else if ( _state == eStages )
  {
    if ( (strcmp( nodeName, kStageTag ) == 0) ||
         (strcmp( nodeName, kPropColTag ) == 0) )
    {
      //  Only enter stage mode if we encounter the xml tag we search for.
      STD_String value;
      if (findAttribute( attributes, kNameTag, value ) && (value.compare(_clipName) == 0))
      {
        _state = eStage;
      }
    }
  }
  else // if ( _state == eStage )
  {
    if ( (strcmp( nodeName, kPlayTag ) == 0) ||
         (strcmp( nodeName, kPropTag ) == 0) )
    {
      TR_NodeTree *nodeTree = new TR_NodeTree;

      STD_String value;
      if (findAttribute(attributes, kSkeletonTag, value))
      {
        XML_ProjectLoader::loadSkeleton(_projectFolder, value, nodeTree);
      }
      if (findAttribute(attributes, kAnimationTag, value))
      {
        XML_ProjectLoader::loadAnimation(_projectFolder, value, nodeTree);
      }
      if (findAttribute(attributes, kDrawingAnimationTag, value))
      {
        XML_ProjectLoader::loadDrawingAnimation(_projectFolder, value, nodeTree);
      }

      STD_String playName;
      findAttribute(attributes, kNameTag, playName);

      //nodeTree->dump();
      _clip->addNodeTree(nodeTree, playName );
    }
    else if ( strcmp( nodeName, kSoundTag ) == 0 )
    {
      float startFrame = 1.0f; // default at first frame.
      STD_String soundName;

      STD_String value;
      if (findAttribute(attributes, kNameTag, value))
      {
        soundName = value;
      }
      if (findAttribute(attributes, kTimeTag, value))
      {
        startFrame = (float)atof(value.c_str());
      }

      _clip->addSoundEvent(soundName, startFrame);
    }
  }

  return this;
}

void XML_StageSaxParser::endComponent(const char *nodeName)
{
  if ( strcmp( nodeName, kStageTag ) == 0 )
  {
    _state = eStages;
  }
  else if ( strcmp( nodeName, kStageColTag ) == 0 )
  {
    _state = eRoot;
  }
}


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
#if 0
#pragma mark - XML_StageNamesSaxParser
#endif

XML_StageNamesSaxParser::XML_StageNamesSaxParser( const STD_String &projectFolder, StringCol_t &clipNames ) :
  _projectFolder(projectFolder),
  _state(eRoot),
  _clipNames(clipNames)
{
}

XML_StageNamesSaxParser::~XML_StageNamesSaxParser()
{
}

XML_SaxParserComponentPtr_t XML_StageNamesSaxParser::startComponent(const char *nodeName, const XML_AttributeCol_t &attributes)
{
  if ( _state == eRoot )
  {
    if ( strcmp( nodeName, kStageColTag ) == 0 )
    {
      _state = eStages;
    }
  }
  else if ( _state == eStages )
  {
    if ( strcmp( nodeName, kStageTag ) == 0 )
    {
      //  Only enter stage mode if we encounter the xml tag we search for.
      STD_String clipName;
      if (findAttribute( attributes, kNameTag, clipName ))
      {
        _clipNames.push_back(clipName);
        _state = eStage;
      }
    }
  }

  return this;
}

void XML_StageNamesSaxParser::endComponent(const char *nodeName)
{
  if ( strcmp( nodeName, kStageTag ) == 0 )
  {
    _state = eStages;
  }
  else if ( strcmp( nodeName, kStageColTag ) == 0 )
  {
    _state = eRoot;
  }
}

File: Project/feather1/Assets/HarmonySDK/Scripts/Xml/XML_StageSaxParser.cs

using System.Collections.Generic;
using System.Globalization;
using System;

using XML_Types;
using XML_SaxParser;
using XML_ProjectLoader;

namespace XML_StageSaxParser
{
  /*!
   *  @class XML_StageNamesSaxParserComponent
   *  Retrieve list of clip names available in project
   */
  public class XML_StageNamesSaxParserComponent : XML_SaxParserComponent
  {
    enum State
    {
      eRoot,
      eStages,
      eStage
    };

    public XML_StageNamesSaxParserComponent()
    {
      _clipNames = new List<string>();
    }

    public override XML_SaxParserComponent startElement( string nodeName, List<XML_Attribute> attributes )
    {
      if ( _state == State.eRoot )
      {
        if ( nodeName.Equals( XML_Constants.kStageColTag ) )
        {
          _state = State.eStages;
        }
      }
      else if ( _state == State.eStages )
      {
        if ( nodeName.Equals( XML_Constants.kStageTag ) )
        {
          //  Only enter stage mode if we encounter the xml tag we search for.
          string clipName = findAttribute( attributes, XML_Constants.kNameTag );
          if ( !string.IsNullOrEmpty(clipName) )
          {
            _clipNames.Add(clipName);
            _state = State.eStage;
          }
        }
      }

      return this;
    }

    public override void endElement( string nodeName )
    {
      if ( nodeName.Equals( XML_Constants.kStageTag ) )
      {
        _state = State.eStages;
      }
      else if ( nodeName.Equals( XML_Constants.kStageColTag ) )
      {
        _state = State.eRoot;
      }
    }

    public List<string> clipNames
    {
      get { return _clipNames; }
    }

    private State        _state;
    private List<string> _clipNames;
  }

  /*!
   *  @class XML_AudioSaxParserComponent
   *  Retrieve list of sound play sequences in clip
   */
  public class XML_AudioSaxParserComponent : XML_SaxParserComponent
  {
    enum State
    {
      eRoot,
      eStages,
      eStage
    };

    public XML_AudioSaxParserComponent( string clipName )
    {
      _clipName = clipName;
      _state = State.eRoot;

      _soundSequences = new List< XML_Types.XML_SoundSequence >();
    }

    public override XML_SaxParserComponent startElement( string nodeName, List<XML_Attribute> attributes )
    {
      if ( _state == State.eRoot )
      {
        if ( nodeName.Equals( XML_Constants.kStageColTag ) )
        {
          _state = State.eStages;
        }
      }
      else if ( _state == State.eStages )
      {
        if ( nodeName.Equals( XML_Constants.kStageTag ) )
        {
          //  Only enter stage mode if we encounter the xml tag we search for.
          string value = findAttribute(attributes, XML_Constants.kNameTag);
          if ( !string.IsNullOrEmpty(value) && value.Equals(_clipName) )
          {
            _state = State.eStage;
          }
        }
      }
      else // if ( _state == eStage )
      {
        if ( nodeName.Equals( XML_Constants.kSoundTag ) )
        {
          XML_Types.XML_SoundSequence seq = new XML_Types.XML_SoundSequence();
          seq._startFrame = 1.0f;

          foreach (XML_Attribute attr in attributes)
          {
            if ( attr._name.Equals( XML_Constants.kNameTag ) )
              seq._name = attr._value;
            else if ( attr._name.Equals( XML_Constants.kTimeTag ) )
              seq._startFrame = float.Parse( attr._value, CultureInfo.InvariantCulture.NumberFormat);
          }

          if ( !string.IsNullOrEmpty(seq._name) )
          {
            _soundSequences.Add(seq);
          }
        }
      }

      return this;

    }

    public override void endElement( string nodeName )
    {
      if ( nodeName.Equals( XML_Constants.kStageTag ) )
      {
        _state = State.eStages;
      }
      else if ( nodeName.Equals( XML_Constants.kStageColTag ) )
      {
        _state = State.eRoot;
      }
    }

    public List< XML_Types.XML_SoundSequence > soundSequences
    {
      get { return _soundSequences; }
    }

    private string                 _clipName;
    private State                  _state;

    private List< XML_Types.XML_SoundSequence >  _soundSequences;
  }

  /*!
   *  @class XML_MetaSaxParserComponent
   *  Retrieve list of meta in clip
   */
  public class XML_MetaSaxParserComponent : XML_SaxParserComponent
  {
    enum State
    {
      eRoot,
      eStages,
      eStage
    };

    public XML_MetaSaxParserComponent( )
    {
      _state     = State.eRoot;

      _props     = new List<XML_PropMeta>();
      _anchors   = new List<XML_AnchorMeta>();
      _metas     = new List<XML_GenericMeta>();
    }

    public override XML_SaxParserComponent startElement( string nodeName, List<XML_Attribute> attributes )
    {
      if ( _state == State.eRoot )
      {
        if ( nodeName.Equals( XML_Constants.kStageColTag ) )
        {
          _state = State.eStages;
        }
      }
      else if ( _state == State.eStages )
      {
        if ( nodeName.Equals( XML_Constants.kStageTag ) ||
             nodeName.Equals( XML_Constants.kPropColTag ) )
        {
          _clipName = findAttribute( attributes, XML_Constants.kNameTag );
          if ( !string.IsNullOrEmpty(_clipName) )
          {
            _state = State.eStage;
          }
        }
      }
      else // if ( _state == eStage )
      {
        if ( nodeName.Equals( XML_Constants.kPropTag ) )
        {
          XML_PropMeta meta;
          meta._playName = findAttribute( attributes, XML_Constants.kNameTag );
          if ( !string.IsNullOrEmpty(meta._playName) )
          {
            meta._clipName = _clipName;

            _props.Add(meta);
          }
        }
        else if ( nodeName.Equals( XML_Constants.kAnchorTag ) )
        {
          XML_AnchorMeta meta;
          meta._nodeName = findAttribute( attributes, XML_Constants.kNodeTag );
          meta._playName = findAttribute( attributes, XML_Constants.kPlayTag );
          if ( !string.IsNullOrEmpty(meta._nodeName) && !string.IsNullOrEmpty(meta._playName) )
          {
            if ( _anchors.IndexOf(meta) < 0 )
            {
              _anchors.Add(meta);
            }
          }
        }
        else if ( nodeName.Equals( XML_Constants.kMetaTag ) )
        {
          XML_GenericMeta meta;
          meta._name = findAttribute( attributes, XML_Constants.kNameTag );
          if ( !string.IsNullOrEmpty(meta._name) )
          {
            meta._clipName = _clipName;
            meta._value = findAttribute( attributes, XML_Constants.kValueTag );
            meta._nodeName = findAttribute( attributes, XML_Constants.kNodeTag );
            meta._playName = findAttribute( attributes, XML_Constants.kPlayTag );

            _metas.Add(meta);
          }
        }
      }

      return this;
    }

    public override void endElement( string nodeName )
    {
      if ( nodeName.Equals( XML_Constants.kStageTag ) )
      {
        _state = State.eStages;
        _clipName = string.Empty;
      }
      else if ( nodeName.Equals( XML_Constants.kStageColTag ) )
      {
        _state = State.eRoot;
      }
    }

    public List<XML_PropMeta> props
    {
      get { return _props; }
    }

    public List<XML_AnchorMeta> anchors
    {
      get { return _anchors; }
    }

    public List<XML_GenericMeta> metas
    {
      get { return _metas; }
    }

    private State                       _state;
    private string                      _clipName;

    private List<XML_PropMeta>          _props;
    private List<XML_AnchorMeta>        _anchors;
    private List<XML_GenericMeta>       _metas;
  }

}
